disk-utils/mkswap.c

```diff
@@ -351,13 +356,17 @@ static unsigned long long get_size(const struct mkswap_control *ctl)
 	int fd;
 	unsigned long long size;
 
-	fd = open(ctl->devname, O_RDONLY);
-	if (fd < 0)
-		err(EXIT_FAILURE, _("cannot open %s"), ctl->devname);
-	if (blkdev_get_size(fd, &size) < 0)
-		err(EXIT_FAILURE, _("cannot determine size of %s"), ctl->devname);
-	if ((unsigned long long) ctl->offset > size)
-		errx(EXIT_FAILURE, _("offset larger than file size"));
+	if (ctl->file && ctl->size)
+		size = ctl->size;
```

Comment:
If the control enters this branch, fd is not initialized.
So close(fd) at line 364/373 causes an unexpected behavior.
A condition guarding the invocation of close needs. e.g.:
```
if (! (ctl->file && ctl->size))
   close(fd);
```

Reply:
I think `fd` can be defined and used only in the `else { }` block, the close(fd) could be at the end of the block.